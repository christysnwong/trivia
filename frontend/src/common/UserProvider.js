import React, {useEffect, useState} from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  stats: {
    level: 14,
    title: "Pro",
    "quizzes completed": 34,
    points: 3440,
    remainingPts: 60,
    levelPts: 300,
  },
};

const UserProvider = ({
  children,
  currUser = demoUser
}) => {
  
  const [currentUser, setCurrUser] = useState();

  return (
    // <UserContext.Provider value={{ currUser }}>
    <UserContext.Provider value={{ currUser, setCurrUser }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
