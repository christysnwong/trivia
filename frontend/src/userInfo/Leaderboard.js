import { useState, useEffect } from "react";
import TriviaApi from "../api/api";

import Table from "../userTable/Table";


const Leaderboard = () => {

    const [gScores, setGScores] = useState([]);

    // fetche leaderboard scores
    useEffect(() => {
      TriviaApi.getLeaderboardScores().then(score => {
          setGScores(score);
      });
    }, []);

    return (
      <div className="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4">

        <h3 className="mb-5">
          <i className="fa-solid fa-globe me-2"></i>
          Leaderboard
        </h3>

        <p className="mb-3">
          The following table shows the top score in each category and
          difficulty. Your score appears in red if you make it to the
          leaderboard (limited to registered users only).
        </p>

        <p className="mb-4">
          Due to the limited number of questions in some categories, each user
          can only have 3 attempts to make a record on the leaderboard.
        </p>

        <Table entries={gScores} leaderboard={true} />
      </div>
    );

}


export default Leaderboard;
