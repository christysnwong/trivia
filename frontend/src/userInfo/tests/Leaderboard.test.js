import React from "react";
import { render, screen } from "@testing-library/react";
import Leaderboard from "../Leaderboard";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../../common/UserProvider";

import TriviaApi from "../../api/api";

const { commonBeforeEach, commonAfterAll } = require("./_testCommonUserInfo");

beforeEach(commonBeforeEach);
afterAll(commonAfterAll);

it("renders without crashing", function () {
  render(
    <UserProvider>
      <Leaderboard />
    </UserProvider>
  );
});

it("matches snapshot", async function () {

  const { asFragment } = await render(
    <MemoryRouter>
      <UserProvider>
        <Leaderboard />
      </UserProvider>
    </MemoryRouter>
  );
  
  expect(asFragment()).toMatchSnapshot();
});

it("checks if results from TriviaApi.getLeaderboardScores are fetched and rendered", async () => {

  jest.spyOn(React, "useEffect").mockImplementation(f => f());
  jest.spyOn(TriviaApi, "getLeaderboardScores");
  render(
    <UserProvider>
      <Leaderboard />
    </UserProvider>
  );

  expect(TriviaApi.getLeaderboardScores).toHaveBeenCalled();

  expect(await screen.findByText(/category/i)).toBeInTheDocument();
  expect(await screen.findByText(/difficulty/i)).toBeInTheDocument();
  expect(await screen.findByText(/points/i)).toBeInTheDocument();
  expect(await screen.findAllByText(/entertainment/i)).toHaveLength(3);
  expect(await screen.findAllByText(/science/i)).toHaveLength(1);
  expect(await screen.findAllByText("testadmin")).toHaveLength(4);


});