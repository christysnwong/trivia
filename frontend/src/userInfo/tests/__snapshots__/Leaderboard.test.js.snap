// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4"
  >
    <h3
      class="mb-5"
    >
      <i
        class="fa-solid fa-globe me-2"
      />
      Leaderboard
    </h3>
    <p
      class="mb-3"
    >
      The following table shows the top score in each category and difficulty. The date and time displayed below are in UTC time zone. Your score appears in red if you make it to the leaderboard (limited to registered users only).
    </p>
    <p
      class="mb-4"
    >
      Due to the limited number of questions in some categories, each user can only have 3 attempts to make a record on the leaderboard.
    </p>
    <div
      class="table-responsive"
    >
      <table
        class="table table-striped table-sm"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              #
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</DocumentFragment>
`;
